k2/torch/csrc/fsa_class.cu

```diff
@@ -0,0 +1,476 @@
+      all_attr_names.insert(name);
+    }
+  }
+  // Check the validation of this fsa, will trigger a fatal error if this fsa
+  // is not valid.
+  Properties();
+
+  // TODO: we also need to pass the name of ragged_labels.
+}
+
+FsaClass::FsaClass(const Ragged<Arc> &fsa, torch::Tensor aux_labels)
+    : fsa(fsa) {
+  K2_CHECK_EQ(fsa.NumElements(), aux_labels.numel());
+  K2_CHECK_EQ(aux_labels.scalar_type(), torch::kInt32);
+  K2_CHECK(ContextFromTensor(aux_labels)->IsCompatible(*fsa.Context()));
+  SetTensorAttr("aux_labels", aux_labels);
+}
+
+FsaClass::FsaClass(const Ragged<Arc> &fsa, Ragged<int32_t> &aux_labels)
+    : fsa(fsa) {
+  K2_CHECK_EQ(fsa.NumElements(), aux_labels.Dim0());
+  K2_CHECK(IsCompatible(fsa, aux_labels));
+  SetRaggedTensorAttr("aux_labels", aux_labels);
+}
+
+FsaClass FsaClass::FromUnaryFunctionTensor(FsaClass &src,
+                                           const Ragged<Arc> &arcs,
+                                           torch::Tensor arc_map) {
+  FsaClass dest(arcs);
+  // Check the validation of the fsa, will trigger a fatal error if the fsa
+  // is not valid.
+  dest.Properties();
+  dest.CopyTensorAttrs(src, arc_map);
+  dest.CopyRaggedTensorAttrs(src, arc_map);
+  return dest;
+}
+
+FsaClass FsaClass::FromUnaryFunctionRagged(FsaClass &src,
+                                           const Ragged<Arc> &arcs,
+                                           Ragged<int32_t> &arc_map) {
+  FsaClass dest(arcs);
+  // Check the validation of the fsa, will trigger a fatal error if the fsa
+  // is not valid.
+  dest.Properties();
+  for (const auto &iter : src.tensor_attrs) {
+    if (iter.second.scalar_type() == torch::kInt32) {
+      torch::Tensor value = iter.second.clone();
+      auto masking =
+          torch::logical_or(torch::ne(src.Labels(), -1), torch::ne(value, -1));
+      // we need a int32_t scalar, so we have to use tensor.
+      auto filler_scalar =
+          torch::tensor(0, torch::dtype(torch::kInt32).device(value.device()));
+      value = torch::where(masking, value, filler_scalar);
+      Array1<int32_t> value_array = Array1FromTorch<int32_t>(value);
+      auto new_value = Index(value_array, arc_map, /*default_value*/ 0);
+      dest.SetRaggedTensorAttr(iter.first, RemoveValuesEq(new_value, 0));
+    } else {
+      K2_CHECK(iter.second.dtype() == torch::kFloat32 ||
+               iter.second.dtype() == torch::kFloat64);
+
+      Dtype dtype = ConvertDtype(iter.second.scalar_type());
+      FOR_REAL_AND_INT32_TYPES(dtype, T, {
+        Array1<T> src_array = Array1FromTorch<T>(iter.second);
+        Ragged<T> s = Index<T>(src_array, arc_map, /*default_value*/ 0);
+        Array1<T> ans_array(s.Context(), s.Dim0());
+        SumPerSublist<T>(s, 0, &ans_array);
+        dest.SetTensorAttr(iter.first, Array1ToTorch(ans_array));
+      });
+    }
+  }
+  dest.CopyRaggedTensorAttrs(src, arc_map);
+  return dest;
+}
+
+FsaClass FsaClass::FromBinaryFunctionTensor(FsaClass &a_src, FsaClass &b_src,
```

Comment:
I think this function will not be used in the decoding.
We can add it later if it is needed.

Just want to make the decoding code as simple as possible.
Please add code that is only needed by decoding.

Reply:
We don't need this function in `Intersect`?

Reply:
No, I don't think we need `k2::Intersect` in decoding.